diff --git a/libosmscout-client-qt/include/osmscoutclientqt/MapWidget.h b/libosmscout-client-qt/include/osmscoutclientqt/MapWidget.h
index ce79300e0..637c60934 100644
--- a/libosmscout-client-qt/include/osmscoutclientqt/MapWidget.h
+++ b/libosmscout-client-qt/include/osmscoutclientqt/MapWidget.h
@@ -204,10 +204,10 @@ public slots:
   void zoomOut(double zoomFactor, const QPoint widgetPosition);
 
   void move(QVector2D vector);
-  void left();
-  void right();
-  void up();
-  void down();
+  void moveLeft();
+  void moveRight();
+  void moveUp();
+  void moveDown();
 
   /**
    * Rotate view to specified angle [radians; [0 ~ 2*PI) ]
diff --git a/libosmscout-client-qt/src/osmscoutclientqt/MapWidget.cpp b/libosmscout-client-qt/src/osmscoutclientqt/MapWidget.cpp
index d032f0e7f..a50e2c186 100644
--- a/libosmscout-client-qt/src/osmscoutclientqt/MapWidget.cpp
+++ b/libosmscout-client-qt/src/osmscoutclientqt/MapWidget.cpp
@@ -469,22 +469,22 @@ void MapWidget::move(QVector2D vector)
     }
 }
 
-void MapWidget::left()
+void MapWidget::moveLeft()
 {
     move(QVector2D( width()/-3, 0 ));
 }
 
-void MapWidget::right()
+void MapWidget::moveRight()
 {
     move(QVector2D( width()/3, 0 ));
 }
 
-void MapWidget::up()
+void MapWidget::moveUp()
 {
     move(QVector2D( 0, height()/-3 ));
 }
 
-void MapWidget::down()
+void MapWidget::moveDown()
 {
     move(QVector2D( 0, height()/+3 ));
 }
