diff --git a/libosmscout-gpx/src/osmscoutgpx/Import.cpp b/libosmscout-gpx/src/osmscoutgpx/Import.cpp
index 0ff091a9e..ea2a4f91e 100644
--- a/libosmscout-gpx/src/osmscoutgpx/Import.cpp
+++ b/libosmscout-gpx/src/osmscoutgpx/Import.cpp
@@ -276,6 +276,7 @@ public:
       return NameSpace::Unknown;
     }
     static std::unordered_map<std::string,NameSpace> nameSpaceMap{
+      {"http://www.topografix.com/GPX/1/0", NameSpace::Gpx},
       {"http://www.topografix.com/GPX/1/1", NameSpace::Gpx},
       {"http://www.garmin.com/xmlschemas/GpxExtensions/v3", NameSpace::GarminExtensions},
     };
